### PR TITLE
HTTP/3: Change alt-svc protocol to h3 and add upgrade test

### LINKED ISSUE #34148: HTTP/3: Test alt-svc upgrade from HTTP/1.1 and HTTP/2 to HTTP/3
Currently blocked from being added to https://github.com/dotnet/aspnetcore/pull/34104

Blocked because HTTP/1.1 and HTTP/2 require suppression of cert error using callback, and HTTP/3 errors when there is a callback.

See https://github.com/dotnet/runtime/issues/55193 and https://github.com/dotnet/runtime/issues/55192

### PR DESCRIPTION
**Pull request by JamesNK**
Fixes https://github.com/dotnet/aspnetcore/issues/34148

### CODE REVIEW

File: src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs [hunk 1]
```diff
@@ -1194,17 +1194,17 @@ private HttpResponseHeaders CreateResponseHeaders(bool appCompleted)
...
+                // TODO: Perf. Avoid allocating enumerator and property's LINQ.
+                // https://github.com/dotnet/aspnetcore/issues/34468
                 foreach (var option in ServerOptions.ListenOptions)
                 {
                     if ((option.Protocols & HttpProtocols.Http3) == HttpProtocols.Http3)
                     {
-                        responseHeaders.HeaderAltSvc = $"h3-29=\":{option.IPEndPoint!.Port}\"; ma=84600";
+                        // TODO: Perf. Create string once instead of per-request.
+                        // https://github.com/dotnet/aspnetcore/issues/34468
+                        responseHeaders.HeaderAltSvc = $"h3=\":{option.IPEndPoint!.Port}\"; ma=84600";
```
Comment (JamesNK) [inline src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs:1207]: Latest draft has finalized "h3"

Comment (Tratcher) [reply src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs:1207]: I was testing the ALPN code with similar values. I think we should include both h3 and h3-29 for compat for at least a milestone, that's really common on websites I've checked. h3 alpn didn't work with HttpClient yet.

Comment (JamesNK) [reply src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs:1207]: I found the opposite with HttpClient: h3-29 doesn’t work and h3 does. That’s why I changed it in this PR.

Comment (JamesNK) [reply src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs:1207]: Chrome (and Edge) support "h3"

image

If HttpClient and major browsers support h3 then I think it's fine to completely switch in RC1.

File: src/Servers/Kestrel/Transport.Quic/test/WebHostTests.cs [hunk 2]
```diff
@@ -152,6 +152,73 @@ public async Task Listen_Http3AndSocketsCoexistOnSameEndpoint_ClientSuccess()
             await host.StopAsync().DefaultTimeout();
         }
 
+        [ConditionalFact]
+        [MsQuicSupported]
+        public async Task Listen_Http3AndSocketsCoexistOnSameEndpoint_AltSvcEnabled_Upgrade()
+        {
+            // Arrange
+            var builder = GetHostBuilder()
```
Comment (Tratcher) [inline src/Servers/Kestrel/Transport.Quic/test/WebHostTests.cs:160]: Doesn't the ALPN setting need to be updated to match?
https://github.com/dotnet/aspnetcore/blob/836d664bee463d7ceb3312df39296151f8b06d0d/src/Servers/Kestrel/Transport.Quic/test/QuicTestHelpers.cs#L28

Comment (JamesNK) [reply src/Servers/Kestrel/Transport.Quic/test/WebHostTests.cs:160]: HttpClient is being inconsistent and using h3 in the altsvc but not as it’s alpn https://github.com/dotnet/runtime/issues/55894?notification_referrer_id=MDE4Ok5vdGlmaWNhdGlvblRocmVhZDIxODY4Nzc0MTY6MzAzMjAx#issuecomment-882839687

Comment (Tratcher) [reply src/Servers/Kestrel/Transport.Quic/test/WebHostTests.cs:160]: Ok. The alt-svc and ALPN should be kept in sync, let's cover this later today.

**Review by Tratcher — APPROVED** [review]

**Pull request merged**